token/services/ttxdb/db/sql/sql.go

```diff
@@ -384,6 +355,118 @@ func (db *Persistence) GetTransactionEndorsementAcks(txID string) (map[string][]
+	now := time.Now().UTC()
+	query := fmt.Sprintf("INSERT INTO %s (token_id, tx_id, tx_index, certification, stored_at) VALUES ($1, $2, $3, $4, $5)", db.table.Certifications)
+
+	tx, err := db.db.Begin()
+	if err != nil {
+		return errors.New("failed starting a transaction")
+	}
+	defer tx.Rollback()
+	for tokenID, certification := range certifications {
+		tokenIDStr := fmt.Sprintf("%s%d", tokenID.TxId, tokenID.Index)
+		logger.Debug(query, tokenIDStr, fmt.Sprintf("(%d bytes)", len(certification)), now)
+		if _, err := tx.Exec(query, tokenIDStr, tokenID.TxId, tokenID.Index, certification, now); err != nil {
+			return errors.Wrapf(err, "failed to execute")
+		}
+	}
+	if err := tx.Commit(); err != nil {
+		return errors.Wrap(err, "failed committing status update")
+	}
+	return nil
+}
+
+func (db *Persistence) ExistsCertification(tokenID *token.ID) bool {
+	tokenIDStr := fmt.Sprintf("%s%d", tokenID.TxId, tokenID.Index)
+	query := fmt.Sprintf("SELECT certification FROM %s WHERE token_id=$1;", db.table.Certifications)
+	logger.Debug(query, tokenIDStr)
+
+	row := db.db.QueryRow(query, tokenIDStr)
+	var certification []byte
+	if err := row.Scan(&certification); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return false
+		}
+		logger.Warnf("tried to check certification existence for token id %s, err %s", tokenIDStr, err)
+		return false
+	}
+	result := len(certification) != 0
+	if !result {
+		logger.Warnf("tried to check certification existence for token id %s, got an empty certification", tokenIDStr)
+	}
+	return result
+}
+
+func (db *Persistence) GetCertifications(ids []*token.ID, callback func(*token.ID, []byte) error) error {
+	if len(ids) == 0 {
+		// nothing to do here
+		return nil
+
+	}
+	// build query
+	conditions, tokenIDs := certificationsQuerySql(ids)
+	query := fmt.Sprintf("SELECT tx_id, tx_index, certification FROM %s WHERE ", db.table.Certifications) + conditions
+
+	rows, err := db.db.Query(query, tokenIDs...)
+	if err != nil {
+		return errors.Wrapf(err, "failed to query")
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var txID string
+		var txIndex int
+		var certification []byte
+		if err := rows.Scan(&txID, &txIndex, &certification); err != nil {
+			return errors.Wrapf(err, "error querying db")
+		}
+		tokenID := &token.ID{
+			TxId:  txID,
+			Index: uint64(txIndex),
+		}
+		if err := callback(tokenID, certification); err != nil {
+			return errors.WithMessagef(err, "failed callback for [%s]", tokenID)
+		}
+	}
+	if err = rows.Err(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// setStatusIfExists checks if the record exists before updating it, because some sql drivers return an
+// error on update of a non-existent record
+func (db *Persistence) setStatusIfExists(tx *sql.Tx, table, txID string, status driver.TxStatus) error {
+	curStatus := driver.Unknown
+	query := fmt.Sprintf("SELECT status FROM %s WHERE tx_id = $1 LIMIT 1;", table)
```

Comment:
Maybe we need to escape characters here. or remember to do it for the future

Reply:
The table name is coming through the tableNames function, which only accepts letters and underscore.